charm/ksql/reactive/ksql.py

```diff
@@ -0,0 +1,234 @@
+
+@hook('stop')
+def uninstall():
+    try:
+        check_call(['snap', 'remove', KSQL_SNAP])
+    except subprocess.CalledProcessError as e:
+        hookenv.log("failed to remove snap: {}".format(e))
+
+def install_snap():
+    check_call(['snap', 'install', '--edge', KSQL_SNAP])
+    set_state('ksql.available')
+
+def upgrade_snap():
+    check_call(['snap', 'refresh', '--edge', KSQL_SNAP])
+    set_state('ksql.available')
+
+@when('ksql.available')
+@when_not('kafka.joined')
+def waiting_for_kafka():
+    ksql = Ksql()
+    ksql.close_ports()
+    ksql.stop()
+    hookenv.status_set('blocked', 'waiting for relation to kafka')
+
+@when_not(
+    'ksql.ca.keystore.saved',
+    'ksql.server.keystore.saved',
+)
+@when('ksql.available')
+def waiting_for_certificates():
+    hookenv.status_set('waiting', 'waiting for easyrsa relation')
+
+@when(
+    'ksql.available',
+    'kafka.ready',
+    'ksql.ca.keystore.saved',
+    'ksql.client.keystore.saved'
+)
+@when_not('ksql.started')
+def configure_ksql(kafka):
+    hookenv.status_set('maintenance', 'setting up ksql')
+
+    ksql = Ksql()
+    ksql.configure(kafka.kafkas())
+    ksql.open_ports()
+    set_state('ksql.started')
+    hookenv.status_set('active', 'ready')
+    # set app version string for juju status output
+    ksql_version = ksql.version() or 'unknown'
+    hookenv.application_version_set(ksql_version)
+
+@when_file_changed(os.path.join(KSQL_SNAP_COMMON, 'etc', "ksql.client.jks"))
+def restart_when_keystore_changed():
+    Ksql().restart()
+
+@when('certificates.available')
+def get_client_cert():
+    # Request client certs from easyrsa
+    tls_client.request_client_cert(
+        'system:snap-ksql-server',
+        crt_path=os.path.join(
+            KSQL_SNAP_COMMON,
+            'etc',
+            'ksql-server.client.crt',
+        ),
+        key_path=os.path.join(
+            KSQL_SNAP_COMMON,
+            'etc',
+            'client.key'
+        )
+    )
+
+@when('tls_client.certs.changed')
+def import_srv_crt_to_keystore():
+    keystore_path = os.path.join(
+        KSQL_SNAP_COMMON,
+        'etc',
+        "ksql.client.jks"
+    )
+    keystore_password = _keystore_password()
+    crt_path = os.path.join(
+        KSQL_SNAP_COMMON,
+        'etc',
+        'ksql-server.client.crt'
+    )
+    key_path = os.path.join(
+        KSQL_SNAP_COMMON,
+        'etc',
+        "client.key"
+    )
+    if os.path.isfile(crt_path) and os.path.isfile(key_path):
+        cert = open(crt_path, 'rt').read()
+        loaded_cert = crypto.load_certificate(
+            crypto.FILETYPE_PEM,
+            cert
+        )
+        key = open(key_path, 'rt').read()
+        loaded_key = crypto.load_privatekey(
+            crypto.FILETYPE_PEM,
+            key
+        )
+
+        cert_changed = data_changed(
+            'ksql_ca_certificate',
+            cert
+        )
+        log('server certificate changed {changed}'.format(changed=cert_changed))
+        if cert_changed:
```

Comment:
I think this check should come first. `if not cert_changed`, we didn't need to read the files above.